articles/WideStrings.md

```diff
@@ -0,0 +1,213 @@
+
+## Multi-byte character background
+
+Before delving into ROS 2 specifics, some time should be spent discussing multi-byte characters in general.
+What is meant by multi-byte characters?
+The original ASCII character set only provided for printable characters using binary sequences 0 to 127.
+Several extensions were made to this to use 128-255, but that approach only extended the character set to be able to cover a few more languages.
+Several other encodings were created as the need arose, and it became clear by the 1990's that a worldwide standard was required.
+Thus, Unicode was created to be a single specification that could cover all of the languages of the world.
+The early versions of Unicode actually failed to live up to this expectation, and some of those early versions of Unicode are still baked into widely used software (Microsoft Windows, Java, etc).
+Later on, the Unicode standard evolved to truly be able to cover the languages of the entire world, and this is what should generally be used for new designs.
+As a side benefit, the newer versions of Unicode (UTF-8) are backwards compatible with ASCII.
+There is a lot of history here, and things are much more complicated then the brief introduction above.
+For more information, the following links provide introductory material:
+
+* [http://kunststube.net/encoding/](http://kunststube.net/encoding/)
+* [http://stackoverflow.com/questions/4588302/why-isnt-wchar-t-widely-used-in-code-for-linux-related-platforms](http://stackoverflow.com/questions/4588302/why-isnt-wchar-t-widely-used-in-code-for-linux-related-platforms)
+* [http://www.diveintopython3.net/strings.html](http://www.diveintopython3.net/strings.html)
+* [http://stackoverflow.com/questions/402283/stdwstring-vs-stdstring](http://stackoverflow.com/questions/402283/stdwstring-vs-stdstring)
+
+The rest of this document is generally going to assume Unicode, unless stated otherwise.
+
+## Unicode Characters in Strings
+ROS 1 says string fields are to contain ASCII encoded data, but allows UTF-8.
+DDS-XTYPES mandates UTF-8 be used as the encoding of IDL type `string`.
+To be compatible with ROS 1 and DDS-XTYPES, in ROS 2 the content of a `string` is expected to be UTF-8.
+
+## Introducing Wide Strings
+ROS 2 messages will have a new [primitive field type](/articles/interface_definition.html) `wstring`.
+This purpose is to allow ROS 2 nodes to comminicate with non-ROS DDS entities using an IDL containing a `wstring` field.
+The encoding of data in this type should be UTF-16 to match DDS-XTYPES 1.2.
+Since both UTF-8 and UTF-16 can encode the same code points, new ROS 2 messages should prefer `string` over `wstring`.
+
+## Encodings are Recommendations
+The choice of UTF-8 or UTF-16 for `string` and `wstring` are recommendations.
+A `string` or `wstring` field is allowed to be populated and published with an unknown encoding.
```

Comment:
The rest of the paragraph reads like "you can do something different but it is probably gone blow". So I am not sure why we should "allow" this. I would rather lean towards being strict and say the encoding shall be UTF-8 / UTF-16. And keep mentioning the technical restrictions from below what they can't contain.

Reply:
In 354da47cb1cf7d9fcc1384e0ac4f6d84f2f1e5a2 encodings are required but not guaranteed to be enforced.